### DISCUSSION TITLE
Bug: unable to get darkmode class in Vue applications with render function

### DISCUSSION THREAD

**Discussion by Mister-Hope** [ArtalkJS/Artalk#367] [General]
I do not think this is expected:

![image](https://user-images.githubusercontent.com/33315834/216306311-f9cfc063-d7c8-4965-9f00-c0c5c671fd33.png)

**Comment by Mister-Hope** [comment]
This should be a bug, but I am not finding the problem is at a glance.

I tried to manually add `atk-dark-mode` class immedialy after alling Artalk.

![image](https://user-images.githubusercontent.com/33315834/216315648-62772a03-33bf-440b-a292-17d20acf26f9.png)

And I can see that the artalk div has  `atk-dark-mode`  class of a few frames then the class is removed. So it seems that the initialization of Artalk is a async process and artalk will choose to "Replace” the dom. I am not finding a api which seems like "onDOMInitialized" at https://github.com/ArtalkJS/Artalk/blob/d85fc4a71ffdc6b151a5d325d135cc2524894790/docs/develop/event.md so I failed to bypass this issue, so I may need some help here.

**Comment by qwqcode** [comment]
The backend provides a new feature called `useBackendConf` that allows the backend API endpoints `/api/conf` or `/api/get` (get comment list) to provide configs to the frontend client, allowing the user to control their frontend config through a graphical user interface.

```
useBackendConf: true (default)

--> user config set in backend
--> when comments list is first loaded (request `/api/get`)
--> config loaded by frontend Artalk client
```

https://artalk.js.org/guide/backend/fe-control.html

By default, `useBackendConf` is enabled and it is recommended that the user sets the frontend config in the GUI, as this makes it easier for the user to configure.

---

The backend API provides the config which can then **override** the frontend `darkMode` config option when the frontend is first loaded.

(i am considering removing the user's control of the `darkMode` in the backend in future releases, as this will break the correct configuration of the frontend)

The dark mode state in Artalk Instance can be set by calling `setDarkMode` after the `conf-loaded` or `list-loaded` event has been triggered (i.e., after the client config provided by the backend API has been loaded).

However, the app lifecycle may be a bit confusing due to the constantly evolving development document and version. I apologise for this. 😅

---

We have implemented a tiny component that integrates with Artalk and is used well in our currently published documentation site.

The dark mode feature works well, as far as I can see, and it may be helpful for reference:

https://github.com/ArtalkJS/Artalk/blob/34e59e653fa6d3fcea9651f02afe67088a3a6d96/docs/.vitepress/theme/Artalk.vue#L67-L80

Related content that may be useful:

https://github.com/ArtalkJS/Artalk/blob/d52b2f52282737f2b1dd1202dea792f71c6efbdc/ui/packages/artalk/src/context.ts#L265

If you have any questions, feel free to reach out to me for further discussion. Thanks!

**Reply by Mister-Hope** [reply]
1.![image](https://user-images.githubusercontent.com/33315834/216350583-c9feb38b-de51-4ef1-bd29-84e58674f066.png)

   I was misleading by this, and I will open a pr to fix this.

2. I do think that it makes no sense to control dark mode at server. A lot of themes allows users to toggle dark mode as they like. If you think this should be a breaking change, then I may still recommend you to shorten the release table or at least warn it in readme.

Also another question to save my time from reading source code: In my opinion imageUploader and avatarUrl (maybe options name are incorrect, but you will know what I am talking) can not be set from server as they are functions, so what'a their behavior? The expected behavior should be they are only allowed to set at client, and won't be effected by useBackendConf, if so I will think that changing dark mode behavior at minor are affordable

**Reply by Mister-Hope** [reply]
BTW The frontend yml link js broken, I am at hospital with my grandma and it's hard for me to view commit log on mobile browser. Maybe you can check and fix that yourself?

**Reply by qwqcode** [reply]
> BTW The frontend yml link js broken, I am at hospital with my grandma and it's hard for me to view commit log on mobile browser. Maybe you can check and fix that yourself?

Wishing your grandmother a quick recovery. I'm sorry for the multiple problems in the documentation. With the merging of three git repositories into a single monorepo, I'll take the necessary action to fix these issues. Thank you for your support and help.

**Reply by Mister-Hope** [reply]
Oh I just have another question which is not mentioned in the docs clearly. It seems that artalk does not bypass local envirnoments.

For example, I would like to have artalk not blocked when debuging with vite devserver, and I can bypass by manually add localhost:port such as `localhost:8080`, however the port may change, and I do perfer to use private ip (e.g.: http://172.26.138.60:8080/) Is there a way for me to do that?

Also, could you please consider bypassing private ip and localhost by default and provide a option for that?

---

Edited: wait, the `*` are supported specially or the field actually support some syntax like glob or regexp?

**Reply by qwqcode** [reply]
I plan to use a keys list to restrict the parameters `imageUploader`, `avatarUrl`, `darkMode` and more similar ones to ensure they **cannot be altered on the backend** and clear documentation should indicate this. Thank you for your response, it is very helpful.

**Reply by qwqcode** [reply]
> Oh I just have another question which is not mentioned in the docs clearly. It seems that artalk does not bypass local envirnoments.
> 
> For example, I would like to have artalk not blocked when debuging with vite devserver, and I can bypass by manually add localhost:port such as `localhost:8080`, however the port may change, and I do perfer to use private ip (e.g.: http://172.26.138.60:8080/) Is there a way for me to do that?
> 
> Also, could you please consider bypassing private ip and localhost by default and provide a option for that?
> 
> Edited: wait, the `*` are supported specially or the field actually support some syntax like glob or regexp?

This is a good point, I was thinking if requests from localhost or 127.0.0.1 should be allowed by default. In the development environment, setting `trusted_domains`[^1] to `["*"]` will turn off cross-domain checks. Support for regular expressions and wildcards is also a good thing, but it increases the complexity of the program. However, I also hope that this feature is available.

Here is some relevant code:

https://github.com/ArtalkJS/Artalk/blob/master/server/middleware/cors/cors.go
https://github.com/ArtalkJS/Artalk/blob/master/server/middleware/site_origin.go#L88
https://github.com/ArtalkJS/Artalk/blob/master/server/middleware/site_origin_test.go

[^1]: https://artalk.js.org/guide/backend/config.html#%E5%8F%AF%E4%BF%A1%E5%9F%9F%E5%90%8D-trusted-domains

**Reply by Mister-Hope** [reply]
So do I need to open a new issue as a feature request for this to remind you?

BTW: I am very unfamiliar about go, my tech stack are mainly at js/ts, with dart, java and a litttle cpp.😅

From the line https://github.com/ArtalkJS/Artalk/blob/0b7b6ca623ebde0555609361dbdd06b3df71c4ea/server/middleware/site_origin.go#L97-L99, it seems that you are supporting `*` specially?

**Reply by qwqcode** [reply]
> So do I need to open a new issue as a feature request for this to remind you?

Yes, a new feature request issue is the way to go. Thanks!

**Comment by Mister-Hope** [comment]
I still need some help here:

```
Artalk.ts:88 TypeError: Cannot read properties of undefined (reading 'appendChild')
    at editor (Artalk.es.js:7411:20)
    at Artalk.es.js:7430:7
    at Array.forEach (<anonymous>)
    at _Artalk.initComponents (Artalk.es.js:7427:32)
    at new _Artalk (Artalk.es.js:7397:10)
    at initArtalk (Artalk.ts:73:18)
```

Artalk is yeilding this, and the line is `this.$root.appendChild(editor.$el);`

Could you plz check your code to see why `this.$root` can be lost during initailization?

**Reply by qwqcode** [reply]
I suspect the issue may be with the `el` config option used to initialize artalk. It's possible that the `el` config is not passing an `HTMLElement` type, or that you're trying to access the element's DOM before the page has finished loading.

To resolve this, consider passing an `HTMLElement` object in the el config and ensuring that it matches `instanceof HTMLElement`. Also, make sure that you're fetching the object at the correct time.

Let me know if there's anything else I can help with or if there are any areas for improvement.

---

Relevant content may helpful:

https://github.com/ArtalkJS/Artalk/blob/d52b2f52282737f2b1dd1202dea792f71c6efbdc/ui/packages/artalk/src/artalk.ts#L46

https://github.com/ArtalkJS/Artalk/blob/d52b2f52282737f2b1dd1202dea792f71c6efbdc/docs/.vitepress/theme/Artalk.vue#L84-L98

https://github.com/ArtalkJS/Artalk/issues/323#issuecomment-1368674659 (or may a HMR issue?)

**Reply by Mister-Hope** [reply]
<img width="545" alt="image" src="https://user-images.githubusercontent.com/33315834/216535824-968b6baa-931d-475f-b3de-226de0c14fd5.png">
<img width="723" alt="image" src="https://user-images.githubusercontent.com/33315834/216535913-c54c7d50-6c5a-481d-a59a-92c58586b9f2.png">

I do think I passed in an element

**Reply by Mister-Hope** [reply]
Also could you explain why `nextTick` is used at line 95?

BTW line 98 is not needed because `false` is the default value for `immediate`

**Reply by Mister-Hope** [reply]
<img width="505" alt="image" src="https://user-images.githubusercontent.com/33315834/216536698-b2b45161-41de-4381-ac11-879ece32e6f3.png">
<img width="499" alt="image" src="https://user-images.githubusercontent.com/33315834/216536714-ad5847f8-b617-496c-8e00-8c1bdf2a33d2.png">
Codes are updated to ensure that artalkContainer has the correct value at the time, so I don;t think the problem should be on my side.

**Reply by qwqcode** [reply]
This issue seems to be related to the app's lifecycle or event listening for navigation routes, as far as I know.

I have previous experience with VuePress, and this is how the issue was resolved, the implementation can be found in the following files:

https://github.com/ArtalkJS/Docs/blob/eef37bca8cc0c9973bf121fdef4014dcd940f104/docs/.vuepress/theme/Artalk.vue

https://github.com/ArtalkJS/Docs/blob/eef37bca8cc0c9973bf121fdef4014dcd940f104/docs/.vuepress/theme/Layout.vue#L3-L8

I initially tried using `nextTick`, but it didn't work consistently. Instead, some documentation frameworks have a built-in `nextTick` that can be used.

In the meantime, I used a workaround of `setTimeout(() => {}, 80)` to instead of `nextTick`.

I am considering ways to improve Artalk's code compatibility with other frameworks such as Vue, VuePress, etc. 🤔

**Reply by Mister-Hope** [reply]
I am not sure but I found an issue at our side at https://github.com/vuepress/vuepress-next/issues/1249 and we may get it fixed soon.

So in my case I am actually calling `Artalk` in promise async twice in the target app.

Is there anything special with cid which artalk may fail when it tries to initializing mutiple instances on a single dom? @qwqcode  E.g.: something like:

```ts
const n = 2; // a value >1
const div = xxx; // a div
const initArtalk = () => {
  Promise.then(()=>{
    // codes initializing artalk on the div
  });
}

// initialize
for(let i=0;i<n;i++){
  initArtalk()
}
```

**Reply by qwqcode** [reply]
> I am not sure but I found an issue at our side at [vuepress/vuepress-next#1249](https://github.com/vuepress/vuepress-next/issues/1249) and we may get it fixed soon.
> 
> So in my case I am actually calling `Artalk` in promise async twice in the target app.
> 
> Is there anything special with cid which artalk may fail when it tries to initializing mutiple instances on a single dom? @qwqcode E.g.: something like:
> 
> ```ts
> const n = 2; // a value >1
> const div = xxx; // a div
> const initArtalk = () => {
>   Promise.then(()=>{
>     // codes initializing artalk on the div
>   });
> }
> 
> // initialize
> for(let i=0;i<n;i++){
>   initArtalk()
> }
> ```

At present, CID (Context ID) is simply a token to determine if it is the same Artalk context and nothing more.

I had previously created a page with two separate comment boxes to allow for multiple Artalk instances to be isolated and not interfere with each other, but it did not prove to be a very useful feature.

**Reply by Mister-Hope** [reply]
> I initially tried using `nextTick`

I am not sure if you understand `nextTick` correctly, vue use time slices to efiiciently update dom with vdom, which means if you are editing values (refs, computed) and they results changes of dom, though they values are updating at once so that computed based on ref and watchers which watch ref and computed can be updated correctly, dom may and may not update during the process, and the update could be happended at any time.

If you need to access the updated dom, `nextTick` is something for that, it will trigger a dom update immediately (i.e.: comparing vdom to dom and update the latter) and trigger callback (or fullfill the promise) after finishing the operation.

So if you are writing code like this:

```vue
<script>
const Artalk = (el)=>{
   // ...

  // artalk access the dom
  if( el.classList.contains('.dark')) {
   // do something
  }
}

const el = ref()
const isDarkmode = ref(false);

const applyDarkmode = () => {
   isDarkmode.value = true;

  // initialize Artalk
  Artalk(el.value)
}
</script>
<template>
  <div key="el" :class="{ dark: isDarkmode }"><div>
</template>
```


When you call `applyDarkmode`, artalk may access a dom which not yet been updated.

So you need to use `nextTick` to ensure this:


```vue
<script>
const Artalk = (el)=>{
   // ...

  // artalk access the dom
  if( el.classList.contains('.dark')) {
   // do something
  }
}

const el = ref()
const isDarkmode = ref(false);

const applyDarkmode = async () => {
   isDarkmode.value = true;


  await  nextTick()

  // we ensure dom has been updated by vue at this time

  // initialize Artalk
  Artalk(el.value)
}
</script>
<template>
  <div key="el" :class="{ dark: isDarkmode }"><div>
</template>
```


So I am not seeing any need for you to use nextTick, as the div is already there when the component is mounted. And it seems that artalk do not care about any atrribute on it.

Also, if developers choose to key the component where Artalk locales, the component dom will not be upadted by vue as long as the key is not changed and any value binded to the component template remain unchanged.

> I am considering ways to improve Artalk's code compatibility with other frameworks such as Vue, VuePress, etc. 🤔

So actaully I think that you should only cares about if the api fits spa. Developers should ensure framework not modifying the dom themselves

**Reply by qwqcode** [reply]
I'm not certain of the specifics of how VuePress works, but my understanding is that switching routes can cause changes to the page data, including elements on the page. If you use Vue's native `nextTick` along with the listeners established by the router, you may not get the correct data, depending on how the app's lifecycle is implemented.

For instance, Taro framework, which is a higher-level wrapper for Vue3, also has a built-in nextTick that operates differently from Vue's. When switching pages, using Taro's nextTick will have a different outcome compared to using Vue's nextTick, ensuring that the code runs correctly after the page switch.

I don't really understand what magic these frameworks use to optimise performance. 😄 However, after the route switch, some data on the page will be changed, whether it is necessary to use nextTick I think is a matter of the program's own life cycle, the most important thing is to ensure that your data can be correctly obtained.

**Reply by qwqcode** [reply]
> So I am not seeing any need for you to use nextTick, as the div is already there when the component is mounted. And it seems that artalk do not care about any atrribute on it.

Since my comment box component (element) is included in the top-level el that will be updated by the routing switch, I have to use `nextTick` or `setTimeout` to make sure I get the element object correctly. I could probably leave the comment box element out of the range of page elements being updated by when the switch is made, and does not use `nextTick`.

**Reply by qwqcode** [reply]
> So actaully I think that you should only cares about if the api fits spa. Developers should ensure framework not modifying the dom themselves

Thanks for your advice. It got me thinking about compatibility issues and inspired me to do the following. I agree with your suggestion to call the `updateConf` and `reload` functions directly in a spa instead of repeatedly creating new instances, as this would reduce the computational overhead and improve performance (although the runtime has a CG). It is true that this is a feature of little use for creating and isolating different instances of Artalk.

**Reply by Mister-Hope** [reply]
Hi, is there a destory method for Artalk? I believed there are memory leaking in your frontend project, as initializing artalk on the same dom repeatly will cause memory increasing during time.

I believed some listeners or local variables/functions may be "shadowed" when covered by a new instance and they won't be GCed

**Reply by qwqcode** [reply]
You have reminded me that there may indeed be a memory leak, but I haven't actually verified it.

In my case, I use the `el.remove()` function to free the dom that artalk is working on, and I use `const` type declarations to avoid this "shadowing" effect.

https://github.com/ArtalkJS/Artalk/blob/34e59e653fa6d3fcea9651f02afe67088a3a6d96/docs/.vitepress/theme/Artalk.vue#L25

https://github.com/ArtalkJS/Artalk/blob/34e59e653fa6d3fcea9651f02afe67088a3a6d96/docs/.vitepress/theme/Artalk.vue#L43

This is really a point to ponder, and I'm not quite sure if there are potential places where it's not CGed. (Although I try to only create listeners associated with one dom of artalk rather than lots of event listeners on the window object)

Anyway, repeatedly creating new is not a good choice, especially in spa, and **I also prefer to abandon the previous initialisation way (and instantiate the object only once)**.

Ensuring that it can be effectively CGed requires some finesse, which also creates difficulty in developing programs that don't leak memory. It is easy to make this mistake.

Concerning the scope of the impact: in a spa page, the program's data is saved continuously, but in a static page (such as a hugo blog), the original data is lost every time the page is refreshed, so it is not affected.
